---
output: github_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
devtools::load_all()
```

# homologene
[![Build Status](https://travis-ci.org/oganm/homologene.svg?branch=master)](https://travis-ci.org/oganm/homologene)[![codecov](https://codecov.io/gh/oganm/homologene/branch/master/graph/badge.svg)](https://codecov.io/gh/oganm/homologene)

An r package that works as a wrapper to homologene

Available species are
* Homo sapiens
* Mus musculus
* Rattus norvegicus
* Danio rerio
* Caenorhabditis elegans
* Drosophila melanogaster
* Rhesus macaque

More species can be added on request

Installation
============
```r
library(devtools)
install_github('oganm/homologene')
```

Usage
===========
Basic homologene function requires a list of genes, and an `inTax` and an `outTax`. In this example, `inTax` is the taxon id of *mus musculus* while `outTax` is for humans.
```{r}
homologene(c('Eno2','Mog'), inTax = 10090, outTax = 9606)
```

For mouse and humans two convenience functions exist that removes the need to provide taxonomic identifiers. Note that the column names are not the same as the `homologene` output.
```{r}
mouse2human(c('Eno2','Mog'))
human2mouse(c('ENO2','MOG','GZMH'))
```
